File: 5. BFS/slick.cpp
#include <bits/stdc++.h>

using namespace std;

const int MAX = 250 + 5;
int dr[] = {-1, 1, 0, 0};
int dc[] = {0, 0, 1, -1};
bool visited[MAX][MAX];
int images[MAX][MAX];
int slickFre[MAX*MAX];
int main(){
  while(1){
    int N, M;
    cin >> N >> M;
    if(N == 0 || M == 0) break;
    // MxN matrix of False values to track visited nodes at [i,j]
    for (int i = 0; i < N; i++){
      for (int j = 0; j < M; j++){
        visited[i][j] = false;
      }
    }
    for (int i = 0; i <= M*N; i++){
        slickFre[i] = 0;
    }
    for (int i = 0; i < N; i++){
      for (int j = 0; j < M; j++){
        cin >> images[i][j];
      }
    }
  
    queue<int> rq;
    queue<int> cq;
    int total = 0;
    for (int i = 0; i < N; i++){
      for (int j = 0; j < M; j++){
        if(visited[i][j]) continue;
        if(!images[i][j]) continue;
        visited[i][j] = true;
        rq.push(i); cq.push(j);
        int move_count = 1;
        while(!rq.empty()){
          int r = rq.front(); rq.pop();
          int c = cq.front(); cq.pop();
          for(int k =0;k<4;k++){
            int rr = r + dr[k];
            int cc = c + dc[k];
            // Skip out of bounds location
            if(rr < 0 || cc < 0) continue;
            if(rr >= N || cc >= M) continue;
            //skip visited location and blocked cells
            if(visited[rr][cc]) continue;
            if(images[rr][cc] == 0) continue;

            move_count ++;
            visited[rr][cc] = true;
            rq.push(rr); cq.push(cc);
          }
        }
        slickFre[move_count]++;
        total ++;
      }
    }
    cout << total << endl;
    for(int i = 0; i<= N*M; i++){
      if(slickFre[i] != 0){
        cout << i << " " << slickFre[i] <<endl;
      }
    }
  }
}

File: 5. BFS/slick.py
import queue
# init direction vectors
dr = [-1, 1, 0, 0]
dc = [0, 0, 1, -1]
MAX = 250
image = [[] for _ in range(MAX)]
while True:
  N, M = map(int, input().strip().split())
  if N == 0 and M == 0: break
  
  slick_fre = [0]*(M*N + 1)
  # MxN matrix of False values to track visited nodes at [i,j]
  visited = [[False for i in range (M)] for _ in range(N)]

  for i in range(N):
    image[i] = list(map(int, input().strip().split()))

  q = queue.Queue()
  total = 0
  for i in range(N):
    for j in range(M):
      if visited[i][j]: continue
      if not image[i][j]: continue
      visited[i][j] = True
      q.put((i,j))
  
      move_count = 1
      while not q.empty():
        (r,c) = q.get()
        for k in range(4):
          rr = r + dr[k]
          cc = c + dc[k]
          # Skip out of bounds location
          if rr < 0 or cc < 0: continue
          if rr >= N or cc >= M: continue
          #skip visited location and blocked cells
          if visited[rr][cc]: continue
          if image[rr][cc] == 0: continue
          move_count +=1
          visited[rr][cc] = True
          q.put((rr,cc))
      slick_fre[move_count] +=1
      total += 1
  print(total)
  for i in range(N*M+1):
    if(slick_fre[i] != 0):
      print(i, slick_fre[i])
